File: sentimental-credit/credit.py
from cs50 import get_int


def main():
    # Prompt for credit card number
    while True:
        card_number = get_int("Number: ")
        if card_number > 0:
            break

    # Check the length of the input
    card_length = len(str(card_number))

    # Check 2 intial number of the credit card
    first_two_digits = int(str(card_number)[:2])
    first_digits = int(str(card_number)[:1])

    # Check length validity
    if card_length not in [13, 15, 16]:
        print("INVALID")
        exit()

    # If card is valid, check which does it belong too
    if luhn_checksum(card_number):
        if card_length == 15 and first_two_digits == 34 or first_two_digits == 37:
            print("AMEX")
        elif card_length == 13 or card_length == 16 and first_digits == 4:
            print("VISA")
        elif card_length == 16 and first_two_digits in [51, 52, 53, 54, 55]:
            print("MASTERCARD")
        else:
            print("INVALID")
    else:
        print("INVALID")

    # function for Luhn's algorithm checksum


def luhn_checksum(card_number):

    # setting up intial variables
    checksum = 0
    card_num = str(card_number)
    digits = len(str(card_num))

    # note: range(start. stop, step)
    for i in range(digits - 1, -1, -2):
        checksum += int(card_num[i])
    for i in range(digits - 2, -1, -2):
        digit = int(card_num[i]) * 2
        checksum += digit // 10 + digit % 10
    return checksum % 10 == 0


main()

File: credit/credit.c
// Luhn's Algorithm Checksum for credit card by Hassan Zakie Yusoff
#include <cs50.h>
#include <stdio.h>

bool luhn_checksum(long card_number) // Luhn's Algorithm checksum
{
    int sum = 0, alternate = false; // setting up intial variables
    while (card_number > 0)
    {
        int digit = card_number % 10; // extract last digit using modulo operator
        card_number /= 10;            // remove last digit using integer division

        if (alternate) // if alternate digit, doubled
        {
            digit *= 2;
            if (digit > 9)
            {
                digit -= 9;
            }
        }

        sum += digit;           // sum all the doubled number
        alternate = !alternate; // togled the alternate as its goes through all the digit
    }

    return sum % 10 == 0; // checked the sum divisibility with 10
}

int main(void)
{
    long card_number;
    do
    {
        card_number = get_long("Number: "); // input of credit card number
    }
    while (card_number < 1); // remove input bellow 1 value and negative value

    long first_2_digits = card_number; // initial value for variable
    int count = 0;
    long number = card_number;

    while (number > 0) // counting the total amount of digit in the credit card number
    {
        count++;
        number = number / 10;
    }

    while (first_2_digits >= 100) // counting the 2 inital number of the credit card
    {
        first_2_digits = first_2_digits / 10;
    }
    // printf("no of digit: %i\n", count); // use if want to check the number of digit counted
    // printf("first two digit number is: %li\n", first_2_digits) // use if want to check the number
    // of first two digit

    if (luhn_checksum(card_number)) // if card is valid, it will check other property
    {
        if (count == 15 && (first_2_digits == 34 || first_2_digits == 37))
        {
            printf("AMEX\n");
        }
        else if (count == 16 && (first_2_digits >= 51 && first_2_digits <= 55))
        {
            printf("MASTERCARD\n");
        }
        else if (first_2_digits /= 10, (count == 13 || count == 16) && first_2_digits == 4)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}
